Add mysql support

laurocaetano opened the pull request:
Continues https://github.com/elixir-lang/ecto/pull/383

**TODO**

- [x] Update documentation
- [x] Add missing `ddl_exists` for `indexes`
- [x] Raise on unsupported types and clauses
    - [x] `returning`
    - [x] `concurrently`
    - [x] `array`
- [x] Fix the `ALTER TABLE` syntax
- [x] Fix casting
- [x] Add unit tests for the unsupported clauses and types
  - [x] `returning`
  - [x] `concurrently`
  - [x] `array`
- [x] Remove the `remove :ok -> {:ok, %{}}` on the `query` function. (Depends on https://github.com/liveforeverx/mariaex/issues/14)
- [x] Change the `insert` and `update` interface to make it return a Keyword list. https://github.com/elixir-lang/ecto/pull/476
- [x] Port https://github.com/elixir-lang/ecto/pull/437
- [x] Port https://github.com/elixir-lang/ecto/pull/411
- [x] Add support for more datatypes on Mariaex
  - [x] datetime/date https://github.com/liveforeverx/mariaex/pull/18
  - [x] boolean https://github.com/liveforeverx/mariaex/pull/21
  - [x] decimal https://github.com/liveforeverx/mariaex/pull/24
  - [x] NULLS https://github.com/liveforeverx/mariaex/pull/28
  - [x] Microseconds https://github.com/liveforeverx/mariaex/pull/30
- [x] Make the missing integration tests pass (Depends on https://github.com/liveforeverx/mariaex/issues/10)
  - [x] migration
  - [x] repo
  - [x] type
  - [x] preload
  - [x] sql_escape
  - [x] transaction
- [x] Fix the `foreign key` clause: https://github.com/elixir-lang/ecto/pull/461/files#diff-975ad8a7b1f8e27ac59595bcdbdd0a07R438
- [x] Verify clauses in SELECT match the order in the planner
- [x] Ensure LAST_INSERT_ID runs in the same DB connection as the insert https://github.com/liveforeverx/mariaex/pull/32

_obs_
  - MySQL doesn't have support to `UUID` type, so it will use `BINARY(16)` instead
  - MySQL doesn't support DDL transactions
  - This PR is work in progress. I will update as I get more stuff done :)

@josevalim thanks for helping :heart:

josevalim commented on integration_test/mysql/test_helper.exs:
```diff
@@ -0,0 +1,60 @@
+{:ok, _pid} = TestRepo.start_link
+{:ok, _pid} = PoolRepo.start_link
+
+# Uncomment when work on adapter MySQL.Connection starts
```
We can remove this comment. :D

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+      end
+    end
+
+    # TODO INSERT INTO POSTS () VALUES();
```
We can remove this comment! :D Or make it about the returning one!

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    end
+
+    defp expr({:in, _, [_left, []]}, _sources) do
+      "false"
```
We need to check if this works for MySQL. If we do `SELECT ... FROM ... WHERE FALSE` does it return no results? For example, Rails was used to do `1=0` here for MySQL for some reason.

laurocaetano replied:
Yes, it does work ->

```SQL
mysql> select * from posts where false;
Empty set (0.00 sec)
```

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    end
+
+    defp expr(list, sources) when is_list(list) do
+      "ARRAY[" <> Enum.map_join(list, ",", &expr(&1, sources)) <> "]"
```
We should probably raise here. List in expressions are not allowed in MySQL...

laurocaetano replied:
Oh, my bad :sweat_smile:

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    end
+
+    defp expr(%Ecto.Query.Tagged{value: other, type: type}, sources) do
+      expr(other, sources) <> "::" <> ecto_to_db(type)
```
I think the casting syntax for MySQL is something like `CAST(... AS ...)`

laurocaetano replied:
Yes, you are right. Thanks

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    end
+
+    defp expr(literal, _sources) when is_float(literal) do
+      String.Chars.Float.to_string(literal) <> "::float"
```
Casting here too!

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    alias Ecto.Migration.Table
+    alias Ecto.Migration.Index
+    alias Ecto.Migration.Reference
+    def ddl_exists(%Table{name: name}) do
```
We are missing the ddl exists for indexes. :)

josevalim replied:
Btw, the reason you didn't add the ddl exists for indexes is because we don't have unit tests for it... because the unit test would be to just copy the SQL posted here... which doesn't make sense. So don't worry... you would have caught it in the integration test anyway. :)

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    defp default_expr(literal) when is_binary(literal),
+      do: "DEFAULT '#{escape_string(literal)}'"
+    defp default_expr(literal) when is_number(literal) or is_boolean(literal),
+      do: "DEFAULT #{literal}" # TODO: Check the boolean here :P
```
Apparently MySQL supports true and false, it just converts them to 0 and 1.

laurocaetano replied:
Yes, that's right.

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,502 @@
+    defp ecto_to_db(:datetime),   do: "timestamp"
+    defp ecto_to_db(:binary),     do: "blob"
+    defp ecto_to_db(:uuid),       do: "binary(16)" # MySQL does not support uuid
+    defp ecto_to_db({:array, _}), do: raise "MySQL doesn't support Array type." # MySQL does not support Array
```
No need to finish exception messages with a dot!

josevalim commented:
Great job! I have added some comments!

laurocaetano commented:
@josevalim, just updated the PR with some of your suggestion. Also I updated the SQL syntax to use the `?` as the query parameter. See: https://github.com/liveforeverx/mariaex/commit/b3e4b53c64c838aa7cfd2f78826d6f87f721f7f2

Now I'm stuck when running the integration tests. The only thing I could track is that it stop working right after the first insert clause.

```elixir
ecto (add-mysql-support) $ MIX_ENV=mysql mix test
** (EXIT from #PID<0.47.0>) {:ecto, :no_connection}
```

I'll try to take a look on it tomorrow. Thanks!

josevalim commented:
@laurocaetano I think this is a timeout issue described here:

https://github.com/liveforeverx/mariaex/issues/11

I will send a pull request to mariaex soon. :)

josevalim commented:
It is not a timeout issue. Basically Mariaex is unable to encode datetime parameters and it is failing hard because of that. It can only encode integers, floats and nulls for now: You could try adding new parameters here:

https://github.com/liveforeverx/mariaex/blob/master/lib/mariaex/messages.ex#L222-L229

You will need to look for MySQL manuals on how to encode them. Ping me if you'd would like to do one of those together!

laurocaetano commented:
@josevalim thanks for the feedback. I haven't had the time to take a look on these issues, but I will spend some time later today. Thanks :)

laurocaetano commented on lib/ecto/adapters/mysql.ex:
```diff
@@ -0,0 +1,113 @@
+  @doc false
+  def insert(repo, source, params, [pk|_], opts) do
+    case super(repo, source, params, [pk], opts) do
+      {:ok, {}} ->
```
cc/ @josevalim

josevalim replied:
:+1:

josevalim commented on lib/ecto/adapters/mysql.ex:
```diff
@@ -0,0 +1,113 @@
+  end
+
+  @doc false
+  def insert(repo, source, params, [pk|_], opts) do
```
Should we raise if we have something else besides the `pk`?

laurocaetano replied:
:+1:

josevalim commented on lib/ecto/adapters/mysql.ex:
```diff
@@ -0,0 +1,113 @@
+      {:ok, {}} ->
+        last_inserted_query = @conn.last_inserted(source, pk)
+        Ecto.Adapters.SQL.model(repo, last_inserted_query, [], opts)
+      err -> err
```
If this can only be an error, you may want to do `{:error, _} = err -> err`.

laurocaetano replied:
:+1:

josevalim commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,519 @@
+    end
+
+    def last_inserted(table, pk) do
+      "SELECT #{quote_name(pk)} FROM #{quote_name(table)} WHERE #{quote_name(pk)} = LAST_INSERT_ID()"
```
As we talked on gtalk, this is not correct, we need to ensure they are in the same connection. Let's revisit this once the update/delete stuff is in place.

laurocaetano replied:
Sure! Just for documentation purpose, this is the documentation for `LAST_INSERT_ID()`: http://dev.mysql.com/doc/refman/5.0/en/information-functions.html#function_last-insert-id

laurocaetano commented on lib/ecto/adapters/mysql/connection.ex:
```diff
@@ -0,0 +1,534 @@
+    end
+
+    defp ecto_to_db(:string),     do: "varchar"
+    defp ecto_to_db(:datetime),   do: "datetime"
```
Floats in Elixir are 64 bit double precision and MySQL uses four bytes for single-precision values (Float) and eight bytes for double-precision (Double) values.

http://elixir-lang.org/getting-started/basic-types.html
http://dev.mysql.com/doc/refman/5.0/en/floating-point-types.html

Shall we save the `Float` type in Elixir as `Double` in MySQL in order to keep precision?

josevalim replied:
No, let's keep them as :float. people can directly increase the precision if they care.

laurocaetano replied:
Ok then, so we might need to change some tests :smile:

laurocaetano commented on integration_test/mysql/all_test.exs:
```diff
@@ -0,0 +1,6 @@
+Code.require_file "../cases/migration.exs", __DIR__
```
Obs:. The `Lock` test case was removed, since MySQL doesn't support it.

josevalim commented on lib/ecto/adapters/mysql.ex:
```diff
@@ -0,0 +1,119 @@
+      env = [{"MYSQL_TCP_PORT", port}|env]
+    end
+
+    args = ["--silent", "-u", database[:username], "-h", database[:hostname], "-e", sql_command]
```
The hostname should be fetched like this: `database[:hostname] || System.get_env("MYSQL_HOST") || "localhost"`

josevalim commented on test/ecto/adapters/mysql_test.exs:
```diff
@@ -0,0 +1,496 @@
+  end
+
+  test "returning" do
+   # assert_raise ArgumentError, "RETURNING is not supported by MySQL", fn ->
```
We can remove this one :)

josevalim commented on lib/ecto/adapters/mysql.ex:
```diff
@@ -0,0 +1,140 @@
+  @doc false
+  def storage_up(opts) do
+    database  = Keyword.fetch!(opts, :database)
+    charset   = Keyword.get(opts, :char_set, "utf8")
```
It should be `:charset`.

josevalim commented on lib/ecto/adapters/mysql.ex:
```diff
@@ -0,0 +1,140 @@
+                           " DEFAULT CHARACTER SET = #{charset} " <> extra
+
+    cond do
+      status == 0                  -> :ok
```
Let's indent this `->`.

ericmj commented:
:cake: :heart:
